### PR TITLE
Check ABI for notifications

### LINKED ISSUE #1879: Check notifications for manifest compliance
**Summary or problem description**
We have events definition as a part of NEP-14 ABI included in the manifest that is deployed along with the contract. It's an important part of contract's interface to the outside world and contract's users may expect that all events emitted would comply with this definition.

But we never check emitted notifications for ABI compliance. One can use an empty list of events in the ABI and then successfully emit whatever he wants to. Or specify some event and then not comply with parameters types specified there. Both can be confusing to any receivers and users of these events.

At the same time lack of any control for this part of ABI provokes developers to not pay any attention to it or even not use it at all which is bad for the network (because it makes manifests less useful).

**Do you have any solution you want to propose?**
Add a check to `RuntimeNotify()` that will reject any notifications not complying with contract's ABI. Failing the execution with some `throw new ArgumentException()` should be perfectly fine for this.

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- Other: contract's ABI and syscalls

### PR DESCRIPTION
**Pull request by erikzhang**
Close #1879

### CODE REVIEW

**Comment by superboyiii** [comment]
A little worried that this will broke old contract and make data incompatible.

**Comment by roman-khimov** [comment]
We need to check mainnet of course, if something breaks this should be a hardfork, but it's a valuable improvement in general.

Now if we could trust NEP-11/NEP-17 manifest specifications (#1883)...

**Comment by vncoelho** [comment]
That can break and may need a hardfork as you mentioned.

However, in a long-term perspective this is a could compliance to be followed by contracts as @roman-khimov emphasized.

**Review by vncoelho — COMMENTED** [review]
The PR is precise @erikzhang ,as always, code looks simple, we will test asap.

**Review by shargon — APPROVED** [review]

**Comment by superboyiii** [comment]
I can't approve this, we need forkheight, otherwise will be break at block 693180 in mainnet. @erikzhang @shargon

**Comment by roman-khimov** [comment]
> otherwise will be break at block 693180 in mainnet

Do you have the hash of the contract causing the problem? We may want to search through the whole chain for these inconsistencies and notify contract developers because they need to upgrade their contracts before the hardfork height.

**Comment by superboyiii** [comment]
> Do you have the hash of the contract causing the problem? We may want to search through the whole chain for these inconsistencies and notify contract developers because they need to upgrade their (broken) contracts before the hardfork height.

`0x08e7a002104fc8f3130e8ac74f550e8a3022e55f` should be from `HumswapBowls`

**Comment by roman-khimov** [comment]
My current list (up to 2215218 height) is:
```
INFO    BAD EVENT       {"contract": "08e7a002104fc8f3130e8ac74f550e8a3022e55f", "event": "Transfer", "error": "parameter 3 type mismatch: ByteArray vs Integer"}
INFO    BAD EVENT       {"contract": "23dcef205da5ed0f48241c8f148bd76dafbc014d", "event": "Transfer", "error": "parameter 3 type mismatch: ByteArray vs Integer"}
INFO    BAD EVENT       {"contract": "42ff28c97a4e77308c2083a7ae4fc1e7fd837364", "event": "OnPausedChanged", "error": "parameter 0 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "4768c475e4c8465f2edf97f265c85950dfebc787", "event": "ApplySwapToPool", "error": "parameter 2 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "4768c475e4c8465f2edf97f265c85950dfebc787", "event": "ComputeSwap", "error": "parameter 2 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "4768c475e4c8465f2edf97f265c85950dfebc787", "event": "OnNEP17Payment", "error": "parameter 3 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "4768c475e4c8465f2edf97f265c85950dfebc787", "event": "Swap", "error": "parameter 3 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "4768c475e4c8465f2edf97f265c85950dfebc787", "event": "SwapFailure", "error": "parameter 3 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "577a51f7d39162c9de1db12a6b319c848e4c54e5", "event": "Debug", "error": "event Debug does not exist"}
INFO    BAD EVENT       {"contract": "65a27d7a6dcdf960bf2992f88b7b70f1057471e3", "event": "Authorized", "error": "parameter 2 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "6e644dda08a62f3fc9d14e824d1a3bd816a0c2d5", "event": "Fault", "error": "mismatch between the number of parameters and items"}
INFO    BAD EVENT       {"contract": "82be94eb34072de6e38b1ec212302c490a998c08", "event": "Authorized", "error": "parameter 2 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "9fa5ffe173dcd8c55dd51ed9720e0d1c9553442c", "event": "Authorized", "error": "parameter 2 type mismatch: Boolean vs Integer"}
INFO    BAD EVENT       {"contract": "cc638d55d99fc81295daccbaf722b84f179fb9c4", "event": "OrderCancelled", "error": "parameter 0 type mismatch: ByteArray vs Integer"}
INFO    BAD EVENT       {"contract": "cc638d55d99fc81295daccbaf722b84f179fb9c4", "event": "OrderCreated", "error": "parameter 1 type mismatch: ByteArray vs Integer"}
INFO    BAD EVENT       {"contract": "cc638d55d99fc81295daccbaf722b84f179fb9c4", "event": "OrderFilled", "error": "parameter 1 type mismatch: ByteArray vs Integer"}
INFO    BAD EVENT       {"contract": "de3fe8d30d98e049272db4b7bd641943b1e86872", "event": "Transfer", "error": "parameter 3 type mismatch: ByteArray vs Integer"}
INFO    BAD EVENT       {"contract": "edf179715646c9f07bd10b52c1eec44407297b8c", "event": "OnPausedChanged", "error": "parameter 0 type mismatch: Boolean vs Integer"}
```
And it's not a small one (please recheck though). As much as I like this change we can't just enable it at some height and hope it works, there will be some breakage and I don't think it's acceptable for mainnet. Suggestions are:
 * for 3.5.0 make this a logged warning instead of a hard failure
 * contact developers of all involved contracts and notify them that this warning will lead to failure in 3.6.0 (with some rough date estimate)
 * allow for some event "normalization", we have a lot of `Boolean vs Integer` for example, but booleans are routinely created by programs with `PUSH0` and `PUSH1` that create integers and it's not a contract developer's fault, so just like we substitute `Buffer` with `ByteString` (for a different reason, but still) we can take an `Integer` and cast it to `Boolean` in the `Notify` itself
 * a bit more controversial, but this can also be done to `Integer` emitted in place of `ByteString`
 * monitor the chain, check contract updates
 * release 3.6.0 with a hardfork that will change warning to failure

**Comment by ixje** [comment]
> My current list (up to 2215218 height) is:

Very useful. I created an issue internally to address the Bool - Integer conversion and then for the contracts created with `boa`  it should be a matter of recompiling with the latest version.

**Comment by ixje** [comment]
> Normalizing in compiler is an option too, even a better one, it's just that it requires contract updates and I'm not sure it can be done in timely manner for mainnet.

I totally agree with that. I was mostly posting with the idea of indicating we're working on having a migration plan ready for the affected contracts created with boa, so the contract author's should to spend as little time as possible on it.

File: src/Neo/SmartContract/ApplicationEngine.Runtime.cs [hunk 1]
```diff
@@ -384,5 +397,47 @@ protected internal void BurnGas(long gas)
...
+            switch (type)
+            {
+                case ContractParameterType.Any:
+                    return true;
+                case ContractParameterType.Boolean:
+                    return aType == StackItemType.Boolean;
+                case ContractParameterType.Integer:
+                    return aType == StackItemType.Integer;
+                case ContractParameterType.ByteArray:
+                case ContractParameterType.String:
```
Comment (roman-khimov) [inline src/Neo/SmartContract/ApplicationEngine.Runtime.cs:414]: Don't we want to have some additional checks for the `String` type? The way it is now it's absolutely the same as `ByteArray`, maybe some differentiation can be useful like if `String` could only contain valid UTF-8 byte sequence.

**Comment by superboyiii** [comment]
@erikzhang Let's move on. It's in 3.5.0 release checklist.

**Comment by superboyiii** [comment]
> * for 3.5.0 make this a logged warning instead of a hard failure

Any example on how to log warning? I'm a little worried if we log it in cli or rpc, still can't be noticed by most people.

**Comment by roman-khimov** [comment]
Well, it's a usual log message (like the one from dBFT or RPC for example), but I know neo-cli does it in a bit different manner than neo-go. We have it in 0.99.4 release already and it's just another log message. Yes, people may not notice it, but they at least will have some chance and it also is tightly related to proper scheduling/announcement of this change. If developers will be aware of the change they will try to take a look at how their contracts behave, if they care about them. If they don't, well, then the contract is obsolete already in some ways.

**Comment by superboyiii** [comment]
> Well, it's a usual log message (like the one from dBFT or RPC for example), but I know neo-cli does it in a bit different manner than neo-go. We have it in 0.99.4 release already and it's just another log message. Yes, people may not notice it, but they at least will have some chance and it also is tightly related to proper scheduling/announcement of this change. If developers will be aware of the change they will try to take a look at how their contracts behave, if they care about them. If they don't, well, then the contract is obsolete already in some ways.

Maybe we can send mail to these contract deployers since their emails are in manifest.

**Comment by superboyiii** [comment]
I've sent a mail to these admins to ask them pay attention for this change, I think we could move on.

**Comment by mfbz** [comment]
Hi all,
I saw from the list that our OnPausedChanged method is triggering a "Boolean vs Integer" error.

This is our event:
`public static event Action<bool> OnPausedChanged;`

This is how it's described in the manifest:
`{"name":"OnPausedChanged","parameters":[{"name":"obj","type":"Boolean"}]}`

This is how we are using it:
```c#
/// <summary>
/// Update contract paused state without access control.
/// </summary>
/// <remarks>
/// Internal function without access restriction.
/// </remarks>
/// <param name="paused">The new paused value to be set.</param>
protected static void UpdatePaused(bool paused)
{
	PausedStore.Put(paused);
	OnPausedChanged(paused);
}`

What are we doing wrong?
Thank you

**Comment by superboyiii** [comment]
> Hi all, I saw from the list that our OnPausedChanged method is triggering a "Boolean vs Integer" error. What are we doing wrong?
> 
> This is our event: `public static event Action<bool> OnPausedChanged;`
> 
> This is how it's described in the manifest: `{"name":"OnPausedChanged","parameters":[{"name":"obj","type":"Boolean"}]}`
> 
> This is how we are using it:
> 
> ```cs
> /// <summary>
> /// Update contract paused state without access control.
> /// </summary>
> /// <remarks>
> /// Internal function without access restriction.
> /// </remarks>
> /// <param name="paused">The new paused value to be set.</param>
> protected static void UpdatePaused(bool paused)
> {
> 	PausedStore.Put(paused);
> 	OnPausedChanged(paused);
> }
> ```

Nothing wrong, currently Neo push integer instead of bool in VM which will cause break in ABI check. It will be fixed in https://github.com/neo-project/neo-vm/pull/497

**Comment by mfbz** [comment]
Ok, so we don't have to change anything?

**Comment by superboyiii** [comment]
> Ok, so we don't have to change anything?

Yes!

**Comment by superboyiii** [comment]
Wait this to be merged in 3.6.0

**Comment by roman-khimov** [comment]
Can we make it logged warning for 3.6.0? If we're to merge it as is it'll break mainnet for sure, but a logged warning may bring a bit more attention to it.

**Comment by shargon** [comment]
Ready to test @superboyiii

File: src/Neo/SmartContract/ApplicationEngine.Runtime.cs (hunk 1 above)
Comment (AnnaShaleva) [reply src/Neo/SmartContract/ApplicationEngine.Runtime.cs:430]: @shargon, how about this additional check?

Comment (shargon) [reply src/Neo/SmartContract/ApplicationEngine.Runtime.cs:430]: @AnnaShaleva  Done

**Review by shargon — DISMISSED** [review]

File: src/Neo/SmartContract/ApplicationEngine.Runtime.cs [hunk 2]
```diff
@@ -428,8 +428,18 @@ private static bool CheckItemType(StackItem item, ContractParameterType type)
                     return aType == StackItemType.Integer;
                 case ContractParameterType.ByteArray:
                 case ContractParameterType.String:
-                    return aType is StackItemType.Any or StackItemType.ByteString or StackItemType.Buffer &&
-                        Utility.StrictUTF8.GetString(item.GetSpan()) is not null; // Prevent any non-UTF8 string
+                    {
+                        if (aType is StackItemType.Any or StackItemType.ByteString or StackItemType.Buffer)
+                        {
```
Comment (AnnaShaleva) [inline src/Neo/SmartContract/ApplicationEngine.Runtime.cs:433]: Now this check is being performed for both cases when `type` is either  `ContractParameterType.ByteArray` or `ContractParameterType.String`. But the initial idea was to perform this check only for `ContractParameterType.String` and let the `ContractParameterType.ByteArray` be as is.

Comment (shargon) [reply src/Neo/SmartContract/ApplicationEngine.Runtime.cs]: My bad :)

File: src/Neo/SmartContract/ApplicationEngine.Runtime.cs [hunk 3]
```diff
@@ -384,5 +413,59 @@ protected internal void BurnGas(long gas)
...
+                            try
+                            {
+                                _ = Utility.StrictUTF8.GetString(item.GetSpan()); // Prevent any non-UTF8 string
+                                return true;
+                            }
+                            catch { }
+                        }
+                        return false;
+                    }
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply src/Neo/SmartContract/ApplicationEngine.Runtime.cs:443]: @AnnaShaleva, you're programming in Go, `string` can't be null there!

Jokes aside, I don't think I have any strong preference here. I'm slightly more in favor of making strings always have some contents (even if it's "", basically the way it is now), but if there are useful cases where `Null` is applicable, we can accept it too. We can also try having more strict policy for now and then relax it if needed (it's always easier than the other way around).

[the remark this answers is not in the document]
Comment (cschuchardt88) [reply src/Neo/SmartContract/ApplicationEngine.Runtime.cs:443]: `StackItemType.Any` normal indicates `NULL`. But normally depends on where it is.
you can always check for `StackItem.IsNull` or `StackItem.Null`
Maybe treat strings the same as `string.IsNullOrEmpty()` does.

**Comment by superboyiii** [comment]
I've compared all block data, at least it's compatible before `HFBasilisk` on both mainnet and testnet. Will make some tests about expected behaviour after `HFBasilisk`.

File: src/Neo/SmartContract/ApplicationEngine.Runtime.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (shargon) [reply src/Neo/SmartContract/ApplicationEngine.Runtime.cs:443]: I think that should be empty, not null

File: src/Neo/SmartContract/ApplicationEngine.Runtime.cs [hunk 4]
```diff
@@ -384,5 +413,59 @@ protected internal void BurnGas(long gas)
...
+                    return true;
+                case ContractParameterType.Boolean:
+                    return aType == StackItemType.Boolean;
+                case ContractParameterType.Integer:
+                    return aType == StackItemType.Integer;
+                case ContractParameterType.ByteArray:
+                    return aType is StackItemType.Any or StackItemType.ByteString or StackItemType.Buffer;
+                case ContractParameterType.String:
+                    {
+                        if (aType is StackItemType.Any or StackItemType.ByteString or StackItemType.Buffer)
```
Comment (AnnaShaleva) [inline src/Neo/SmartContract/ApplicationEngine.Runtime.cs:433]:
```suggestion
                        if (aType is StackItemType.ByteString or StackItemType.Buffer)
```

Comment (AnnaShaleva) [reply src/Neo/SmartContract/ApplicationEngine.Runtime.cs:433]: Just to make it more explicit.

File: src/Neo/SmartContract/ApplicationEngine.Runtime.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply src/Neo/SmartContract/ApplicationEngine.Runtime.cs:443]: @shargon, good, then see the https://github.com/neo-project/neo/pull/2810#discussion_r1296990236, please.

**Review by steven1227 — APPROVED** [review]

**Review by Liaojinghui — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]

**Review by Liaojinghui — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Comment by cschuchardt88** [comment]
```csharp
// is
Runtime.Notify("Hello", new[] { "World" });
// the as
[DisplayName("Hello")]
public static event Action<string> event_name;
event_name("world");
```

Just wondering if the above calls the same function in the VM. I don't make smart contracts so i don't know.

**Pull request merged**